desc/particles.py

```diff
@@ -0,0 +1,1248 @@
+        diffrax Solver object to use in integration. Defaults to
+        `Tsit5(scan_kind='bounded')`, a RK45 explicit solver.
+    adjoint : diffrax.AbstractAdjoint, optional
+        How to take derivatives of the trajectories. `RecursiveCheckpointAdjoint`
+        supports reverse mode AD and tends to be the most efficient. For forward mode AD
+        use `diffrax.ForwardMode()`.
+    chunk_size : int, optional
+        Chunk size for integration over particles. If None (default), the integration
+        will be done over all particles at once without chunking.
+    options : dict, optional
+        Additional keyword arguments to pass to the field computation,
+            - iota : Profile
+                Iota profile of the Equilibrium, if not already assigned.
+            - source_grid: Grid
+                Source grid to use for field computation during Biot-Savart.
+    throw : bool, optional
+        Whether to throw an error if the integration fails. If False, will return NaN
+        for the points where the integration failed. Defaults to True.
+    return_aux : bool, optional
+        Whether to return auxiliary information from the integrator. If True, will
+        return a tuple (x, v, aux) where aux consists `ts`, `stats` and `result`
+        from `diffrax.diffeqsolve`. Defaults to False. `ts` may become useful if
+        `SaveAt(steps=True)` is used (see `_trace_particles`). Note that `x`, `v` and
+        `ts` will be padded with NaNs to `max_steps` size in that case.
+
+    Returns
+    -------
+    x : ndarray, shape(num_particles, num_timesteps, 3)
+        Position of each particle at each requested time, in
+        either r,phi,z or rho,theta,zeta depending ``model.frame``.
+    v : ndarray
+        Velocity of each particle at specified times. The exact number of columns
+        will depend on ``model.vcoords``.
+
+    """
+    if not params:
+        params = field.params_dict
+    if not options:
+        options = {}
+
+    if bounds is None:
+        bounds = jnp.array([[0, jnp.inf], [-jnp.inf, jnp.inf], [-jnp.inf, jnp.inf]])
+        if isinstance(field, Equilibrium):
+            bounds = bounds.at[0, 1].set(1.0)  # rho bounds for flux coordinates
+
+    def default_event(t, y, args, **kwargs):
+        if isinstance(field, Equilibrium):
+            i = jnp.sqrt(y[0] ** 2 + y[1] ** 2)
+            j = jnp.arctan2(y[1], y[0])
+        else:
+            i = y[0]
+            j = y[1]
+        i_out = jnp.logical_or(i < bounds[0, 0], i > bounds[0, 1])
+        j_out = jnp.logical_or(j < bounds[1, 0], j > bounds[1, 1])
+        k_out = jnp.logical_or(y[2] < bounds[2, 0], y[2] > bounds[2, 1])
+        return jnp.logical_or(i_out, jnp.logical_or(j_out, k_out))
+
+    stepsize_controller = PIDController(
+        rtol=rtol, atol=atol, dtmin=min_step_size, pcoeff=0.3, icoeff=0.3, dcoeff=0
+    )
+    max_steps = setdefault(max_steps, int((ts[-1] - ts[0]) / min_step_size * 10))
```

Comment:
I don't think we need the factor of 10 here. Supposing every step had dt=dtmin, then the total number of steps would be `(ts[-1] - ts[0]) / dtmin`, which would be the maximum possible (since more steps would require dt < dtmin)